R/functions__eager.R

```diff
@@ -37,41 +38,88 @@
+    parallel = TRUE,
+    # eager = FALSE,
+    to_supertypes = FALSE) {
+  if (exists("do_browse", .GlobalEnv) && do_browse) browser()
+
+  # unpack arg list
+  l = unpack_list(..., skip_classes = "data.frame")
+
   ## Check inputs
-  how = match.arg(how[1L], c("vertical", "horizontal", "diagonal"))
+  how_args = c("vertical", "horizontal", "diagonal") # , "vertical_relaxed", "diangonal_relaxed")
+
+  how = match.arg(how[1L], how_args) |>
+    result() |>
+    unwrap("in pl$concat()")
 
-  # dispatch on item class and how
   first = l[[1L]]
-  result = pcase(
-    inherits(first, "DataFrame"),
+  eager = !inherits(first, "LazyFrame")
+  args_modified = names(as.list(sys.call()[-1L]))
+
+  # dispatch on item class and how
+
+  Result_out = pcase(
+    how == "vertical" && (inherits(first, "Series") || is.vector(first)),
     {
-      vdf = l_to_vdf(l)
-      pcase(
-        how == "vertical",   concat_df(vdf),
-        how == "diagonal",   diag_concat_df(vdf),
-        how == "horizontal", hor_concat_df(vdf),
-        or_else = stopf("Internal error")
-      )
+      if (any(args_modified %in% c("parallel"))) {
+        warning(
+          "in pl:concat(): args: parallel takes no effect when concatenating Series",
+          call. = FALSE
+        )
+      }
+      concat_series(l, rechunk, to_supertypes)
+    },
+    how == "vertical",
+    concat_lf(l, rechunk, parallel, to_supertypes),
+    how == "diagonal",
+    {
+      if (any(args_modified %in% c("to_supertypes"))) {
+        warning(
+          "Args to_supertypes",
+          "takes no effect for how=='diagonal'",
+          call. = FALSE
+        )
+      }
+      diag_concat_lf(l, rechunk, parallel)
     },
-    inherits(first, "Series"),
+    how == "horizontal" && !eager,
     {
-      stopf("not implemented Series")
+      Err_plain(
+        "how=='horizontal' is not supported for lazy (first element is LazyFrame).",
+        "Try e.g. <LazyFrame>$join() to get Lazy join or pl$concat(lf1$collect(),lf2,lf3).",
+        "to get a eager horizontal concatenation"
+      )
     },
-    inherits(first, "Expr"),
+    how == "horizontal",
     {
-      stopf("not implemented Expr")
+      if (any(args_modified %in% c("parallel", "to_supertypes"))) {
+        warning(
+          "Args parallel, rechunk, eager and to_supertypes",
+          "takes no effect for how=='horizontal'",
+          call. = FALSE
+        )
+      }
+      hor_concat_df(l)
     },
 
     # TODO implement Series, Expr, Lazy etc
-    or_else = stopf(paste0("type of first list element: '", class(first), "' is not supported"))
+    or_else = Err_plain("internal error:", how, "not handled")
   )
 
-  unwrap(result)
+  # convert back from lazy if eager
+  and_then(Result_out, \(x) {
+    pcase(
+      inherits(x, "DataFrame") && !eager, Err_plain("internal logical error in pl$concat()"),
+      inherits(x, "LazyFrame") && eager, Ok(x$collect()),
```

Comment:
This shouldn't be possible (related to py-polars error above)

Reply:
it is just a runtime assertion, to avoid a future silent bug in some refactor

Reply:
I have added a comment why it still needs to be there